src/main/cpp/include/genomicsdb/genomicsdb_columnar_field.h

```diff
@@ -0,0 +1,326 @@
+      m_buffer.resize(num_bytes);
+      m_valid.resize(num_bytes);
+      m_num_live_entries = 0ull;
+      m_num_filled_entries = 0ull;
+      m_num_unprocessed_entries = 0ull;
+      m_next_buffer = 0;
+      m_previous_buffer = 0;
+    }
+    //Delete both move and copy constructors
+    GenomicsDBBuffer(const GenomicsDBBuffer& other) = delete;
+    GenomicsDBBuffer(GenomicsDBBuffer&& other) = delete;
+    //Functions
+    inline void set_is_in_live_list(const bool val) { m_is_in_live_list = val; }
+    inline bool is_in_live_list() const { return m_is_in_live_list; }
+    //Live and filled entry control
+    inline size_t get_num_live_entries() const { return m_num_live_entries; }
+    inline void set_num_live_entries(const size_t n)
+    {
+      m_num_live_entries = n;
+      m_num_filled_entries = n;
+      m_num_unprocessed_entries = n;
+    }
+    inline void decrement_num_live_entries(const size_t value)
+    {
+      assert(m_num_live_entries >= value);
+      m_num_live_entries -= value;
+    }
+    inline void increment_num_live_entries(const size_t val=1u)
+    {
+      m_num_live_entries += val;
+    }
+    inline size_t get_num_unprocessed_entries() const
+    {
+      return m_num_unprocessed_entries;
+    }
+    inline void decrement_num_unprocessed_entries(const size_t value)
+    {
+      assert(m_num_unprocessed_entries >= value);
+      m_num_unprocessed_entries -= value;
+    }
+    inline std::vector<uint8_t>& get_buffer() { return m_buffer; }
+    inline const std::vector<uint8_t>& get_buffer() const { return m_buffer; }
+    inline uint8_t* get_buffer_pointer() { return &(m_buffer[0]); }
+    inline const uint8_t* get_buffer_pointer() const { return &(m_buffer[0]); }
+    inline size_t get_buffer_size_in_bytes() const { return m_buffer.size(); }
+    /*inline std::vector<size_t>& get_offsets() { return m_offsets; }*/
+    inline size_t* get_offsets_pointer() { return &(m_offsets[0]); }
+    inline size_t get_offsets_size_in_bytes() const { return ((m_offsets.size()-1u)*sizeof(size_t)); } //last offsets element holds the "size"
+    inline size_t get_offsets_length() const { return (m_offsets.size()-1u); } //last offsets element holds the "size"
+    /*
+     * Set offset in buffer of the idx-th element
+     * Useful only for variable length fields and used by the iterators
+     * This allows us to use m_offsets for computing the size of every cell data without
+     * having to write "if(last_cell) else" expressions
+     */
+    inline void set_offset(const size_t idx, const size_t value)
+    {
+      assert(idx < m_offsets.size());
+      m_offsets[idx] = value;
+    }
+    inline size_t get_offset(const size_t idx) const
+    {
+      assert(idx < m_offsets.size() && idx < m_num_filled_entries);
+      return m_offsets[idx ];
+    }
+    inline size_t get_size_of_variable_length_field(const size_t idx) const
+    {
+      assert(idx+1u < m_offsets.size() && idx < m_num_filled_entries);
```

Comment:
If the buffer fits N elements, TIleDB returns offsets for N elements and a total size. The GenomicsDBBuffer structure allocates N+1 offsets elements, uses N for the offsets and stores  the total size in the N+1th offset element. This way all the sizes are computed using offsets without any if(last_cell) statements